packages/react-components/lib/tasks/task-timeline.tsx

```diff
@@ -0,0 +1,122 @@
+import * as RmfModels from 'rmf-models';
+import { rosTimeToJs } from '../utils';
+
+const getPhaseColors = (theme: Theme) => ({
+  pending: theme.palette.info.light,
+  completed: theme.palette.success.light,
+  failed: theme.palette.error.light,
+});
+
+const useStyles = makeStyles((theme) => {
+  const phaseColors = getPhaseColors(theme);
+  return {
+    paper: {
+      padding: '6px 16px',
+    },
+    secondaryTail: {
+      backgroundColor: theme.palette.secondary.main,
+    },
+    pendingPhase: {
+      background: phaseColors.pending,
+    },
+    completedPhase: {
+      background: phaseColors.completed,
+    },
+    failedPhase: {
+      background: phaseColors.failed,
+    },
+    dotInfo: {
+      textOverflow: 'ellipsis',
+      overflow: 'hidden',
+    },
+  };
+});
+
+export interface TaskTimelineProps {
+  taskSummary: RmfModels.TaskSummary;
+}
+
+export function TaskTimeline({ taskSummary }: TaskTimelineProps): JSX.Element {
+  const classes = useStyles();
+  const theme = useTheme();
+  const phaseColors = getPhaseColors(theme);
+  const timelinePhases = taskSummary.status.split('\n\n');
+  const currentDotIdx = timelinePhases.findIndex((msg) => msg.startsWith('*'));
+  const timelineInfo = taskSummary.status.split('\n\n');
+
+  const timelineDotProps = timelinePhases.map((_, idx) => {
+    if (
+      [RmfModels.TaskSummary.STATE_CANCELED, RmfModels.TaskSummary.STATE_FAILED].includes(
+        taskSummary.state,
+      )
+    ) {
+      return {
+        className: classes.failedPhase,
+        color: phaseColors.failed,
+      };
+    }
+
+    if (taskSummary.state === RmfModels.TaskSummary.STATE_COMPLETED) {
+      return {
+        className: classes.completedPhase,
+        color: phaseColors.completed,
+      };
+    }
+
+    if (taskSummary.state === RmfModels.TaskSummary.STATE_ACTIVE && idx < currentDotIdx) {
+      return {
+        className: classes.completedPhase,
+        color: phaseColors.completed,
+      };
+    }
+
+    return {
+      className: classes.pendingPhase,
+      color: phaseColors.pending,
+    };
+  });
+
+  return (
+    <Timeline>
+      {timelineInfo.map((dotInfo, idx) => {
+        return (
+          <TimelineItem key={idx}>
+            <TimelineOppositeContent>
+              <Typography variant="body2" color="textSecondary">
+                {idx == 0 && rosTimeToJs(taskSummary.start_time).toLocaleTimeString()}
+                {idx != 0 &&
+                  idx == timelineInfo.length - 1 &&
+                  rosTimeToJs(taskSummary.end_time).toLocaleTimeString()}
+              </Typography>
+            </TimelineOppositeContent>
+            <TimelineSeparator>
+              <TimelineDot className={timelineDotProps[idx].className} />
+              {idx < timelineInfo.length - 1 && <TimelineConnector />}
+              {idx == timelineInfo.length - 1 && (
+                <TimelineConnector style={{ background: 'transparent' }} />
+              )}
```

Comment:
Why not just skip rendering the last connector?

Reply:
I think when I was implementing this, I had a situation where removing the last connector will shift the last dot and info component towards the left if the connector is not rendered, but that seems to not happen now, so I will remove this additional connector. might have missed it previously

Reply:
:+1: feel free to add this in if that problem comes back, maybe with a comment explaining why. Just want to make sure this is intended and not an oversight.